dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java

```diff
@@ -553,14 +555,21 @@ private int createCommand(CommandType commandType, int processDefineId,
                         }
                     }
                     if (!CollectionUtils.isEmpty(listDate)) {
-                        // loop by schedule date
-                        for (Date date : listDate) {
-                            cmdParam.put(CMDPARAM_COMPLEMENT_DATA_START_DATE, DateUtils.dateToString(date));
-                            cmdParam.put(CMDPARAM_COMPLEMENT_DATA_END_DATE, DateUtils.dateToString(date));
+
+                        int effectThreadsCount = expectedParallelismNumber == null ? 1 : Math.min(listDate.size(), expectedParallelismNumber);
+                        logger.info("In parallel mode, current expectedParallelismNumber:{}", expectedParallelismNumber);
+
+                        int average = listDate.size() / effectThreadsCount;
+                        int slice = listDate.size() % effectThreadsCount == 0 ? average : average + 1;
+
+                        Lists.partition(listDate, slice).stream().forEach(partition -> {
+                            cmdParam.put(CMDPARAM_COMPLEMENT_DATA_START_DATE, DateUtils.dateToString(partition.get(0)));
+                            cmdParam.put(CMDPARAM_COMPLEMENT_DATA_END_DATE, DateUtils.dateToString(partition.get(partition.size() - 1)));
```

Comment:
With the current change, I think the last mission cannot be executed.

```
  List<Schedule> schedules = processService.queryReleaseSchedulerListByProcessDefinitionId(processDefinitionId);
        List<Date> listDate = Lists.newLinkedList();
        if (!CollectionUtils.isEmpty(schedules)) {
            for (Schedule schedule : schedules) {
                listDate.addAll(CronUtils.getSelfFireDateList(startDate, endDate, schedule.getCrontab()));
            }
        }
        // get first fire date
        Iterator<Date> iterator = null;
        Date scheduleDate;
        if (!CollectionUtils.isEmpty(listDate)) {
            iterator = listDate.iterator();
            scheduleDate = iterator.next();
            processInstance.setScheduleTime(scheduleDate);
            processService.updateProcessInstance(processInstance);
        } else {
            scheduleDate = processInstance.getScheduleTime();
            if (scheduleDate == null) {
                scheduleDate = startDate;
            }
        }
```
[executeComplementProcess](https://github.com/apache/dolphinscheduler/blob/dev/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java)
```
  public static List<Date> getSelfFireDateList(Date startTime, Date endTime, CronExpression cronExpression) {
    List<Date> dateList = new ArrayList<>();

    while (Stopper.isRunning()) {
      startTime = cronExpression.getNextValidTimeAfter(startTime);
      if (startTime.after(endTime) || startTime.equals(endTime)) {
        break;
      }
      dateList.add(startTime);
    }

    return dateList;
  }
```
[getSelfFireDateList](https://github.com/apache/dolphinscheduler/blob/dev/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/quartz/cron/CronUtils.java)

I think the condition judgment of startTime.equals(endTime) should be removed in the getSelfFireDateList method

Reply:
Thanks @zhuangchong.
In fact, when startTime is same as endTime, the listDate would be empty, and then the code would run into the  [ExecutorServiceImpl.java#L567](https://github.com/apache/dolphinscheduler/blob/f5675170311ccc3cc8e6f88e74c956220297e8cd/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java#L567) like below:
```java
else {
                        // loop by day
                        int runCunt = 0;
                        while (!start.after(end)) {
                            runCunt += 1;
                            cmdParam.put(CMDPARAM_COMPLEMENT_DATA_START_DATE, DateUtils.dateToString(start));
                            cmdParam.put(CMDPARAM_COMPLEMENT_DATA_END_DATE, DateUtils.dateToString(start));
                            command.setCommandParam(JSONUtils.toJsonString(cmdParam));
                            processService.createCommand(command);
                            start = DateUtils.getSomeDay(start, 1);
                        }
                        return runCunt;
                    }
```
the runCunt would be 1.
The part I updated will not affect this case.

Reply:
@kyoty 

I feel that we are not talking about a part.

What you describe is the judgment operation of the first complement generation task

What I am expressing is the execution operation after generating the complement task,
In the complement parallel task, multiple serial tasks are generated according to the custom parallelism. Each serial task has a start time and an end time (the start time of the nth time). In this serial task, when the loop is the nth time If the start time is equal to the end time, it will not be executed this time.

My understanding may be wrong ...

Reply:
@zhuangchong Sorry for the delay.

This part of the code is very complicated.

First of all, we must know that: **`CronUtils#getSelfFireDateList` will only execute with the time points that satisfy cronExpression in the open time interval (startTime, endTime) .**

Then there are currently two complements, all needs to be taken into consideration:

1. Serial complement, `CronUtils#getSelfFireDateList` will only be executed once in `MasterExecThread#executeComplementProcess`.

2. Parallel complement, `CronUtils#getSelfFireDateList` will be executed twice: `ExecutorServiceImpl#createCommand` and `MasterExecThread#executeComplementProcess`, **executed twice will cause the `startTime` and `endTime` of each prepared-execute timeInterval in custom-parallel mode to be discarded**.

**I dare not modify `CronUtils.getSelfFireDateList`, because modifying it will also affect the serial complement case.**

 **In order to be compatible with all scenarios at the same time, I modified the time range split in parallel mode, such as the prepared-execute timeInterval is [A,B], but we would split it into  [A-1, B+1].  In this way, after executing `CronUtils.getSelfFireDateList` twice, we would finally got the correct interval.**

Now I've implemented it as I described above, and the manual test works well.

Do you have any ieda?